---
title: "Class 19: Pertussis and the CMI-PB project"
author: "Jiachen Fan (A17662703)"
format: pdf
---

# 1. Investigating pertussis cases by year

We can view this data on the CDC website here: https://www.cdc.gov/pertussis/surv-reporting/cases-by-year.html
> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
# install.packages("datapasta")
```

```{r}
#| echo: False
cdc <- data.frame(
                                 year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L,2020L,2021L),
         cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617,
                                          6124,2116)
       )
```

```{r}
library(tidyverse)
```


```{r}
library(ggplot2)

ggplot(cdc) +
  aes(x= year,y= cases) +
  geom_point() +
  geom_line() +
  labs(title = "Pertusis Cases by Year (1922-2019)")
```
# 2. A tale of two vaccines (wP & aP)
> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
ggplot(cdc) +
  aes(x= year,y= cases) +
  geom_point() +
  geom_line() +
  geom_vline(xintercept=1946, color ="red") +
  geom_vline(xintercept=1996, color ="blue") +
  geom_vline(xintercept=2019, color ="darkgreen")+
  labs(title = "Pertusis Cases by Year (1922-2019)")
```
Reported cases reduced significantly after introduction of wP vaccine.

> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

 1) more sensitive PCR-based testing, 2) vaccination hesitancy 3) bacterial evolution (escape from vaccine immunity) 4) waning of immunity in adolescents originally primed as infants with the newer aP vaccine as compared to the older wP vaccine.
 
Additional points for discussion: How are vaccines currently approved?

Typically we first examine ‘Correlates of protection’ which are things that can be measured within weeks or months after vaccination, and which are thought to correlate with increased protection from disease. For the aP vaccine this was an induction of antibodies against pertussis toxin (PT) in infants at equivalent levels to those induced by the wP vaccine. The aP vaccines also had less side effects (reduction of sore arms, fever and pain).
Testing for protection induced by a new vaccine requires a lot of people exposed to the pathogen (like in a pandemic).
It is impossible to discover a effect 10 years post vaccination in the current trial system.
It is unclear what differentiates people that have been primed with aP vs. wP long term.
The CMI-PB project is an attempt to make data on this question open and examinable by all.

# 3. Exploring CMI-PB data

```{r}
library(jsonlite)
```


```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
```

```{r}
head(subject, 3)
```

> Q4. How many aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```
There are 60 aP and 58 wP in the dataset.

> Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```
There are 79 Female and 39 Male patients in the dataset.

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$biological_sex, subject$race)
```
> Q. Make a histogram of the subject age distribution and facet by infancy_vac


```{r}
library(lubridate)
```
What is today’s date
```{r}
today()
```

```{r}
subject$age <- time_length(ymd(subject$date_of_boost)-ymd(subject$year_of_birth), "years")
```

```{r}
ggplot(subject)+
  aes(age,
      fill = infancy_vac)+
  facet_wrap(vars(infancy_vac),nrow = 2)+
  geom_histogram()
```
> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

```{r}
subject$age_today <- time_length(today()-ymd(subject$year_of_birth),"years")
round(mean(subject$age_today[subject$infancy_vac=="wP"]))
```
```{r}
round(mean(subject$age_today[subject$infancy_vac=="aP"]))
```
```{r}
t.test(subject$age_today[subject$infancy_vac=="wP"],subject$age_today[subject$infancy_vac=="aP"])$p.v
```
i) 36; ii) 26; iii) Yes

```{r}
table(subject$dataset)
```
Joining multiple tables

```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/v4/plasma_ab_titer", simplifyVector = TRUE) 
```

```{r}
head(specimen)
```

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
library(dplyr)

meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```
> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
head(abdata)
```
> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

> Q12. What are the different $dataset values in abdata and what do you notice about the number of rows for the most “recent” dataset?

```{r}
table(abdata$dataset)
```
# 4. Examine IgG Ab titer levels
```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens:

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen) +
  geom_boxplot() 
    
```
```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen) +
  geom_boxplot() + 
  xlim(0,75) +
  facet_wrap(vars(visit), nrow=2)
```
> Q14. What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others?

PT,PRN,FIM2/3,FHA show differences. These are surface molecules of pertussis that the patients are vaccinated against.

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()
```
```{r}
igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```
> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“OVA”, that is not in our vaccines) and a clear antigen of interest (“PT”, Pertussis Toxin, one of the key virulence factors produced by the bacterium B. pertussis).

```{r}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```
```{r}
filter(igg, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = TRUE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```
> Q16. What do you notice about these two antigens time courses and the PT data in particular?

PT levels clearly rise over time but not OVA. This trend is similar between wP and aP subjects.

> Q17. Do you see any clear difference in aP vs. wP responses?

No

```{r}
oops <- abdata %>% filter(antigen =="Fim2/3")
table(oops$dataset)
```
I want a time course of IgG MFI_normalised
```{r}
#abdata$planned_day_relative_to_boost
```
```{r}
igpt.21 <- abdata %>%filter(dataset == "2021_dataset",
                            isotype == "IgG",
                            antigen == "PT")
```


```{r}
ggplot(igpt.21) +
  aes(planned_day_relative_to_boost, 
      MFI_normalised, 
      col=infancy_vac) +
  geom_point(alpha=0.6) +
  geom_line(aes(group=subject_id),linewidth=0.5,alpha=0.5) +
  geom_smooth(se= FALSE, span =0.4,linewidth=3)
```


```{r}
igpt.22 <- abdata %>%filter(dataset == "2022_dataset",
                            isotype == "IgG",
                            antigen == "PT")
ggplot(igpt.22) +
  aes(planned_day_relative_to_boost, 
      MFI_normalised, 
      col=infancy_vac) +
  geom_point(alpha=0.6) +
  geom_line(aes(group=subject_id),linewidth=0.5,alpha=0.5) +
  geom_smooth(se= FALSE, span =0.4,linewidth=3)
```

> Q18. Does this trend look similar for the 2020 dataset?

```{r}
igpt.20 <- abdata %>%filter(dataset == "2020_dataset",
                            isotype == "IgG",
                            antigen == "PT")
ggplot(igpt.20) +
  aes(planned_day_relative_to_boost, 
      MFI_normalised, 
      col=infancy_vac) +
  geom_point(alpha=0.6) +
  geom_line(aes(group=subject_id),linewidth=0.5,alpha=0.5) +
  geom_smooth(se= FALSE, span =0.4,linewidth=3)
```

No.

# 5. Obtaining CMI-PB RNASeq data

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 


ssrna <- inner_join(rna, meta)
```

> Q19. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```
> Q20.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

The maximum level is at the 4th visit.

> Q21. Does this pattern in time match the trend of antibody titer data? If not, why not?

No.

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

